### PR TITLE
Nemotron perf changes

### PR DESCRIPTION
**Pull request by b8zhong**
## TODOs:
This PR aims to address the very obvious issues. There are still the following items:
1. Ensure the right kernels
- When using the default command, it uses Flashinfer FA2 attention for full attention, instead we should use mamba `extra_buffer` and `trtllm_mha` for that command (always). For example, we'd use:
```
python3 -m sglang.launch_server \
  --model-path nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16 \
  --trust-remote-code \
  --attention-backend trtllm_mha \
  --max-mamba-cache-size 256 \
  --mem-fraction-static 0.8 \
  --mamba-scheduler-strategy extra_buffer \
  --tp 4 \
  --ep 4 \
  --mamba-backend flashinfer
```
2. Kernel perf
- For Mamba SSU, Flashinfer provides a kernel for this (`--mamba-backend flashinfer`) It's faster in all scenarios, but does not support topk > 1 in spec decode.

## Motivation

The numbers are for BS = 1
Item 1: https://github.com/sgl-project/sglang/pull/26622 (105->127.67 TPS) (this PR includes https://github.com/sgl-project/sglang/actions/runs/26609018294)

Item 2: Change the router GEMM from FP32 x FP32 to BF16 with FP32 output. (127.67 TPS -> 130.61 TPS)

Before:
Screenshot 2026-05-29 at 8 55 58 AM

After:
Screenshot 2026-05-29 at 8 36 07 AM

```
python3 -m sglang.test.run_eval --port 30000 --eval-name gpqa --num-examples 198 --num-threads 198 --repeat 8 --max-tokens 131072 --temperature 1.0 --top-p 0.95 --chat-template-kwargs '{"enable_thinking": true}'
BEFORE:
127.67 TPS
Repeat: 8, mean: 0.811
Scores: ['0.798', '0.838', '0.803', '0.803', '0.823', '0.813', '0.813', '0.793']
Mean latency: 6915.596 s

AFTER:
130.61 TPS
Repeat: 8, mean: 0.809
Scores: ['0.783', '0.823', '0.803', '0.798', '0.828', '0.828', '0.783', '0.823']
Mean latency: 5995.849 s
```

Item 3: Add Mamba2 to _layer_norm_fwd_1pass_kernel (130.61 -> 147.48 TPS)
This kernel supports num_groups > 1, allow it. It will fuse 10 elementwise
Before:
Screenshot 2026-05-29 at 8 30 00 AM
After:
Screenshot 2026-05-29 at 8 30 43 AM

Item 4: Run Relu2 with new JIT CUDA kernel. (147.48 TPS -> 148 TPS) (number is low, because it's already on aux stream, and when the BS is low it doesn't block the main MoE stream)
Before:
Screenshot 2026-05-29 at 8 28 03 AM
After:
Screenshot 2026-05-29 at 8 28 56 AM

Item 5: Elimenate the routed SF mul, and do it in TRTLLM kernel. (around 150 TPS)
Before:
Screenshot 2026-05-29 at 9 52 25 AM

After:
Screenshot 2026-05-29 at 9 54 28 AM

Item 6: Eliminate the weird memcpy, which is from the following code (151.21 TPS):
```python3
output[:num_actual_tokens], _ = self.out_proj(hidden_states)
```
Before:
Screenshot 2026-05-29 at 10 14 06 AM

After:
Screenshot 2026-05-29 at 10 14 55 AM

## Accuracy Tests

Final GPQA accuracy results:
```
python3 -m sglang.launch_server \
    --model-path nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16 \
    --trust-remote-code \
    --tool-call-parser qwen3_coder \
    --reasoning-parser nemotron_3 --attention-backend=flashinfer --tp 4 --ep 4 --max-running-requests 512
```

```
Repeat: 8, mean: 0.812███████████████████████████████████████████████████████████████████████████████▌                                                                                                  | 98/198 [1:34:35<58:52, 35.33s/it]
Scores: ['0.813', '0.823', '0.808', '0.823', '0.833', '0.823', '0.788', '0.788']
Mean latency: 5676.146 s
```

With trtllm_mha, extra buffer, and Flashinfer SSU:
```
python3 -m sglang.launch_server \
  --model-path nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16 \
  --trust-remote-code \
  --attention-backend trtllm_mha \
  --max-mamba-cache-size 256 \
  --tool-call-parser qwen3_coder \
  --reasoning-parser nemotron_3 \
  --mamba-scheduler-strategy extra_buffer \
  --tp 4 \
  --ep 4 \
  --mamba-backend flashinfer --max-running-requests 512 --max-mamba-cache-size 2560
```

Since we didn't change any code of trtllm_mha or Flashinfer SSU, the little accuracy degredation of 2% is not from this PR (it's also unexpected..) we should investigate why.
```
Repeat: 8, mean: 0.787
Scores: ['0.758', '0.783', '0.788', '0.773', '0.763', '0.823', '0.803', '0.808']███████████████████████████████████████████████████████████████████████████████▊                                       | 158/198 [1:21:08<14:07, 21.18s/it]
Mean latency: 4709.060 s
```

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #26905712683](https://github.com/sgl-project/sglang/actions/runs/26905712683)
Latest PR Test (Extra): :white_check_mark: [Run #27045144634](https://github.com/sgl-project/sglang/actions/runs/27045144634)

### CODE REVIEW

**Comment by b8zhong** [comment]
/rerun-test test/registered/models_e2e/test_nvidia_nemotron_3_nano.py

**Comment by b8zhong** [comment]
/rerun-test test/registered/models_e2e/test_nvidia_nemotron_3_nano.py test/registered/models_e2e/test_nvidia_nemotron_3_super_bf16.py test/registered/models_e2e/test_nvidia_nemotron_3_super_bf16_mtp.py
  test/registered/4-gpu-models/test_nvidia_nemotron_3_super_nvfp4.py test/registered/8-gpu-models/test_nvidia_nemotron_3_super_nightly.py test/registered/lora/test_lora_nemotron_3_super_120b_a12b_logprob_diff.py
  test/registered/unit/models/test_nemotron_h_weight_loading.py

**Comment by b8zhong** [comment]
/rerun-test test/registered/models_e2e/test_nvidia_nemotron_3_nano.py test/registered/models_e2e/test_nvidia_nemotron_3_super_bf16.py test/registered/models_e2e/test_nvidia_nemotron_3_super_bf16_mtp.py
test/registered/4-gpu-models/test_nvidia_nemotron_3_super_nvfp4.py test/registered/8-gpu-models/test_nvidia_nemotron_3_super_nightly.py test/registered/lora/test_lora_nemotron_3_super_120b_a12b_logprob_diff.py
test/registered/unit/models/test_nemotron_h_weight_loading.py

File: python/sglang/srt/models/nemotron_h.py [hunk 1]
```diff
@@ -280,15 +287,10 @@ def _forward_core_shared_routed_overlap(
 
     def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         num_tokens, hidden_dim = hidden_states.shape
+        # routed_scaling_factor is fused into the experts call (applied by the
+        # MoE runner / topk), so final_hidden_states is already scaled.
         final_hidden_states, shared_output = self._forward_core(hidden_states)
 
-        # Fix FP16 overflow
```
Comment (b8zhong) [inline python/sglang/srt/models/nemotron_h.py:285]: Note: this is intentional. We can do the multiply inside the fused computation, where it operates on fp32 values. And the current Nemotron tests can pass this

**Comment by b8zhong** [comment]
/rerun-test test_nvidia_nemotron_3_super_bf16.py test_nvidia_nemotron_3_nano.py

File: python/sglang/srt/arg_groups/nemotron_h_hook.py [hunk 2]
```diff
@@ -29,28 +30,30 @@ def apply_nemotron_h_defaults(server_args: "ServerArgs", model_arch: str) -> Non
...
-            ):
-                server_args.moe_runner_backend = "marlin"
-                logger.info(
-                    "Use marlin as MoE runner backend on SM80-SM90 for "
-                    f"{model_arch} {model_config.quantization}"
-                )
-            else:
-                server_args.moe_runner_backend = "flashinfer_cutlass"
+
+    if (is_modelopt or model_config.quantization is None) and (
```
Comment (Fridge003) [inline python/sglang/srt/arg_groups/nemotron_h_hook.py:34]: Why model_config.quantization could be None here

File: python/sglang/jit_kernel/csrc/elementwise/activation.cuh [hunk 3]
```diff
@@ -173,6 +201,50 @@ struct ActivationKernel {
     RuntimeCheck(expert_step >= 1, "expert_step must be positive");
     launch(input, out, type, static_cast<const int32_t*>(expert_ids.data_ptr()), static_cast<uint32_t>(expert_step));
   }
+
+  template <ActivationKind kAct>
+  static constexpr auto unary_kernel = act_kernel<T, kAct, kUsePDL>;
```
Comment (Fridge003) [inline python/sglang/jit_kernel/csrc/elementwise/activation.cuh:206]: Where is kUsePDL controlled?

File: python/sglang/srt/layers/attention/mamba/mamba.py [hunk 4]
```diff
@@ -718,9 +722,11 @@ def forward(
         hidden_states = self.norm(preallocated_ssm_out, gate[:num_actual_tokens])
 
         # 5. Final linear projection
-        output[:num_actual_tokens], _ = self.out_proj(hidden_states)
+        mixer_out, _ = self.out_proj(hidden_states)
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/mamba/mamba.py:725]: Why doing this

File: python/sglang/srt/layers/attention/mamba/mixer2_rms_norm_gated.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def forward_cuda(
             # Keep gate in float32 for numerical stability during silu
             return x * torch.nn.functional.silu(gate.to(torch.float32)).to(input_dtype)
 
-        if ((self.n_groups % self.tp_size) != 0) or self.n_groups != 1:
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/mamba/mixer2_rms_norm_gated.py:109]: Why changing this, will there be risk

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 6]
```diff
@@ -711,15 +711,15 @@ def forward(
...
+        output: Optional[torch.Tensor],
         layer_id: int,
         forward_batch: ForwardBatch,
         mup_vector: Optional[torch.Tensor] = None,
         use_triton_causal_conv: bool = False,
     ):
         assert isinstance(self.forward_metadata, Mamba2Metadata)
         layer_cache = self.req_to_token_pool.mamba2_layer_cache(layer_id)
-        intermediate_states = mixer.forward(
+        mixer_out, intermediate_states = mixer.forward(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:722]: Why doing this

File: python/sglang/srt/layers/quantization/unquant.py [hunk 7]
```diff
@@ -509,6 +510,13 @@ def forward_cuda(
                     else ActivationType.Swiglu
                 ),
             )[0]
+
+            if (
+                not layer.should_fuse_routed_scaling_factor_in_topk
```
Comment (Fridge003) [inline python/sglang/srt/layers/quantization/unquant.py:515]: Is it only happening on nemotron model?

File: python/sglang/srt/arg_groups/nemotron_h_hook.py (hunk 2 above)
Comment (b8zhong) [reply python/sglang/srt/arg_groups/nemotron_h_hook.py:34]: It means unquantized (like BF16). I also agree, it's not very clear..

File: python/sglang/jit_kernel/csrc/elementwise/activation.cuh (hunk 3 above)
Comment (b8zhong) [reply python/sglang/jit_kernel/csrc/elementwise/activation.cuh:206]: It will be controlled by `make_cpp_args(dtype, is_arch_support_pdl())` (template at build time, instead of runtime arg)

File: python/sglang/srt/layers/quantization/unquant.py (hunk 7 above)
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/unquant.py:515]: No. When the backend/topk does not fold the scaling factor in, the quant method applies it on the output instead (to avoid double apply of routed SF). So it should just be like the original behaviour

**Comment by nvpohanh** [comment]
> For Mamba SSU, Flashinfer provides a kernel for this (--mamba-backend flashinfer) It's faster in all scenarios, but does not support topk > 1 in spec decode.

Sounds like a feature request we can make for FlashInfer team?

**Comment by b8zhong** [comment]
@nvpohanh , do you know if it's a common use case? (Not sure if it's used in SA Qwen 3.5, etc)

**Comment by Fridge003** [comment]
@b8zhong Also please fix this failure
https://github.com/sgl-project/sglang/actions/runs/26903997048/job/79363730421

**Comment by b8zhong** [comment]
/rerun-test test_nvidia_nemotron_3_super_bf16.py test_nvidia_nemotron_3_nano.py
